kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java

```diff
@@ -58,7 +60,33 @@ public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
                 LOGGER.debug("{}: Dispatching downstream {} request to filter{}: {}",
                         ctx.channel(), decodedFrame.apiKey(), filterDescriptor(), msg);
             }
-            invoker.onRequest(decodedFrame.apiKey(), decodedFrame.apiVersion(), decodedFrame.header(), decodedFrame.body(), filterContext);
+
+            var stage = invoker.onRequest(decodedFrame.apiKey(), decodedFrame.apiVersion(), decodedFrame.header(),
+                    decodedFrame.body(), filterContext);
+            stage.whenComplete((filterResult, t) -> {
```

Comment:
To prevent out-of-order issues, the logic here will need to look like this:

- if stage is completed already, just process the as you see now (that's the fast path)
- if state is still pending
   - stop netty reading from the client by toggling autoread
   - queue up any requests that are between downstream and this filter.
   - when the stage completes, process it
   - process any queued up requests
   - toggle autoread to restart the flow from the client.